src/System.ObjectModel/src/System/Collections/ObjectModel/ObservableCollection.cs

```diff
@@ -121,11 +124,104 @@ protected override void RemoveItem(int index)
 
             base.RemoveItem(index);
 
+            if (!_skipRaisingEvents)
+            {
+                OnCountPropertyChanged();
+                OnIndexerPropertyChanged();
+                OnCollectionChanged(NotifyCollectionChangedAction.Remove, removedItem, index);
+            }
+        }
+
+        /// <summary>
+        /// Called by base class Collection&lt;T&gt; when a count of items is removed from the list;
+        /// raises a CollectionChanged event to any listeners.
+        /// </summary>
+        protected override void RemoveItemsRange(int index, int count)
+        {
+            CheckReentrancy();
+
+            int length = this.Count;
+            if (index < this.Count - count)
+            {
+                length = index + count;
+            }
+
+            T[] oldItems = new T[Items.Count];
+            if (Items is List<T> list)
+            {
+                list.CopyTo(oldItems);
+            }
+            else
+            {
+                for (int i = 0; i < oldItems.Length; i++)
+                {
+                    oldItems[i] = this[i];
+                }
+            }
```

Comment:
Why do we need to copy the entire collection to a temporary array? Why not something like:

```c#
protected override void RemoveItemsRange(int index, int count)
{
    CheckReentrancy();

    T[] removedItems = null;

    bool ignore = _skipRaisingEvents;
    if (!ignore)
    {
        _skipRaisingEvents = true;

        if (count > 0)
        {
            removedItems = new T[count];
            for (int i = 0; i < count; i++)
            {
                removedItems[i] = this[index + i];
            }
        }
    }

    try
    {
        base.RemoveItemsRange(index, count);
    }
    finally
    {
        if (!ignore)
        {
            _skipRaisingEvents = false;
        }
    }

    if (count > 0 && !_skipRaisingEvents)
    {
        OnCountPropertyChanged();
        OnIndexerPropertyChanged();
        OnCollectionChanged(NotifyCollectionChangedAction.Remove, removedItems, index);
    }
}
```

Reply:
This is a much different way than I was approaching the problem. I love how it only requires 1 array. I have used this implementation and all of the tests are passing. Thanks!!